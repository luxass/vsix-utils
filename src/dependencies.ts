import type { Manifest, PackageManager } from "./types";
import { exec } from "node:child_process";
import path, { isAbsolute, join, resolve } from "node:path";
import process from "node:process";
import { promisify } from "node:util";

const execAsync = promisify(exec);

export interface ExtensionDependenciesOptions {
  /**
   * The package manager to use.
   */
  packageManager: PackageManager;

  /**
   * The current working directory
   * @default process.cwd()
   */
  cwd: string;
}

export interface ExtensionDependency {
  /**
   * The name of the dependency.
   */
  name: string;

  /**
   * The version of the dependency.
   */
  version?: string;

  /**
   * The path to the dependency.
   */
  path: string;
}

interface PnpmDependency {
  from: string;
  version: string;
  resolved: string;
  path: string;
  dependencies: Record<string, PnpmDependency>;
}

interface YarnTreeNode {
  name: string;
  children: YarnTreeNode[];
}

interface YarnDependency {
  name: string;
  path: string;
  version: string;
  children: YarnDependency[];
}

/**
 * Retrieves all production dependencies for an extension based on the package manager being used.
 *
 * @param {Manifest} manifest - The extension's manifest object containing dependency information
 * @param {ExtensionDependenciesOptions} options - Configuration options for retrieving dependencies
 *
 * @returns {Promise<ExtensionDependenciesResult} Promise resolving to an object containing:
 * - dependencies: Array of extension dependencies with name, version and path
 * - packageManager: The package manager that was used
 *
 * @throws Error if:
 * - Package manager cannot be detected when using 'auto'
 * - Unsupported package manager is detected/specified (e.g. deno, bun)
 * - Unable to parse dependency information from package manager output
 *
 * @remarks
 * Supports npm, yarn, and pnpm package managers.
 * When using npm, parses output of `npm list`
 * When using yarn, parses output of `yarn list`
 * When using pnpm, parses output of `pnpm list`
 */
export async function getExtensionDependencies(manifest: Manifest, options: ExtensionDependenciesOptions): Promise<ExtensionDependency[]> {
  const {
    packageManager,
    cwd = process.cwd(),
  } = options;

  const dependencies = new Set<ExtensionDependency>();

  if (packageManager === "npm") {
    const { stdout } = await execAsync("npm list --production --parseable --depth=99999 --loglevel=error", { cwd });
    const lines = stdout.split(/[\r\n]/).filter((path) => isAbsolute(path));

    for (const line of lines) {
      if (line === resolve(cwd)) {
        continue;
      }

      const dependency = line.split(`${path.sep}node_modules${path.sep}`)[1];

      if (dependency == null) {
        throw new Error(`could not parse dependency: ${line}`);
      }

      dependencies.add({
        name: dependency.replace(/\\/g, "/"),
        version: manifest.dependencies != null ? manifest.dependencies[dependency.replace(/\\/g, "/")] : undefined,
        path: line,
      });
    }
  } else if (packageManager === "yarn") {
    const { stdout } = await execAsync("yarn list --prod --json", { cwd });

    const match = /^\{"type":"tree".*$/m.exec(stdout);

    if (!match || match.length !== 1) {
      throw new Error("Could not parse result of `yarn list --json`");
    }

    const trees = JSON.parse(match[0]).data.trees as YarnTreeNode[];

    if (!Array.isArray(trees) || trees.length === 0) {
      return [];
    }

    const prune = true; // TODO: using packaged dependencies

    function asYarnDependency(prefix: string, tree: YarnTreeNode, prune: boolean): YarnDependency | null {
      if (prune && /@[\^~]/.test(tree.name)) {
        return null;
      }

      let name: string;
      let version: string = "";
      try {
        const tmp = tree.name.split("@");

        if (tmp[0] === "") {
          tmp.shift();
          tmp[0] = `@${tmp[0]}`;
        }

        name = tmp[0]!;
        version = tmp[1] || "";
      } catch {
        name = tree.name.replace(/^([^@+])@.*$/, "$1");
      }

      const dependencyPath = path.join(prefix, name);
      const children: YarnDependency[] = [];

      for (const child of tree.children || []) {
        const dep = asYarnDependency(path.join(prefix, name, "node_modules"), child, prune);

        if (dep) {
          children.push(dep);
        }
      }

      return { name, path: dependencyPath, children, version };
    }

    const result = trees.map((tree) => asYarnDependency(join(cwd, "node_modules"), tree, prune)).filter((dep) => dep != null);

    const internalDeps = new Set<string>();

    const flatten = (dep: YarnDependency) => {
      if (internalDeps.has(dep.path)) {
        return;
      }

      dependencies.add({
        name: dep.name,
        version: dep.version,
        path: dep.path,
      });
      internalDeps.add(dep.path);

      if (dep.children) {
        for (const child of dep.children) {
          flatten(child);
        }
      }
    };

    for (const dep of result) {
      flatten(dep);
    }
  } else if (packageManager === "pnpm") {
    // use --ignore-workspace to avoid always including the workspace packages
    const { stdout } = await execAsync("pnpm list --production --json --depth=99999 --loglevel=error --ignore-workspace", { cwd });
    let entryList = [];
    try {
      entryList = JSON.parse(stdout);
    } catch {
      return [];
    }

    if (!Array.isArray(entryList) || entryList.length === 0) {
      return [];
    }

    const entry = entryList[0] as {
      dependencies?: Record<string, PnpmDependency>;
    };

    if (entry == null || typeof entry !== "object" || entry.dependencies == null || typeof entry.dependencies !== "object") {
      return [];
    }

    const internalDeps = new Set<string>();

    const flatten = (dep: PnpmDependency) => {
      if (internalDeps.has(dep.path)) {
        return;
      }

      dependencies.add({
        name: dep.from,
        version: dep.version,
        path: dep.path,
      });
      internalDeps.add(dep.path);

      if (dep.dependencies) {
        for (const child of Object.values(dep.dependencies)) {
          flatten(child);
        }
      }
    };

    for (const value of Object.values(entry.dependencies)) {
      flatten(value);
    }
  } else {
    throw new Error(`unsupported package manager: ${packageManager}`);
  }

  return Array.from(dependencies);
}
