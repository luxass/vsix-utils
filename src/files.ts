/**
 * @module files
 *
 * This module contains utility functions for handling files.
 */

import type { Buffer } from "node:buffer";
import type { ExtensionDependency } from "./dependencies";
import type { ManifestAsset } from "./manifest";
import type { Manifest, PackageManager } from "./types";
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path, { extname } from "node:path";
import process from "node:process";
import ignore from "ignore";
import mime from "mime";
import { detect } from "package-manager-detector";
import { glob } from "tinyglobby";
import { VSCE_DEFAULT_IGNORE } from "./vsce-constants";

export interface VsixLocalFile {
  type: "local";
  path: string;
  readonly localPath: string;
}

export interface VsixInMemoryFile {
  type: "in-memory";
  path: string;
  readonly contents: Buffer | string;
}

export type VsixFile = VsixLocalFile | VsixInMemoryFile;

export function isLocalFile(file: VsixFile): file is VsixLocalFile {
  return file.type === "local";
}

export function isInMemoryFile(file: VsixFile): file is VsixInMemoryFile {
  return file.type === "in-memory";
}

export interface CollectOptions {
  /**
   * The directory where the extension is located.
   * @default process.cwd()
   */
  cwd?: string;

  /**
   * The file to use for ignoring files.
   */
  ignoreFile?: string;

  /**
   * The dependencies to include in the package.
   */
  dependencies?: ExtensionDependency[];

  /**
   * README file path
   * @default "README.md"
   */
  readme?: string;
}

/**
 * Collects files for a VSIX package based on the provided manifest and options.
 *
 * @param {Manifest} manifest - The VSIX manifest containing package details
 * @param {CollectOptions} options - Configuration options for file collection
 * @param {string?} options.cwd - The current working directory (defaults to process.cwd())
 * @param {string?} options.ignoreFile - The name of the ignore file to use (defaults to ".vscodeignore")
 * @param {ExtensionDependency[]?} options.dependencies - The dependencies to include in the package
 * @param {string?} options.readme - The name of the readme file to include (defaults to "README.md")
 *
 * @returns {Promise<VsixFile[]>} A promise that resolves to an array of VsixFile objects representing the files to be included
 * in the VSIX package. Each file object contains the local path and the target path in the extension.
 *
 * @remarks
 * The function takes into account both .gitignore and .vscodeignore files if they exist.
 * It filters files based on the ignore patterns and includes necessary files like package.json
 * and the readme file.
 */
export async function collect(manifest: Manifest, options: CollectOptions): Promise<VsixFile[]> {
  const {
    cwd = process.cwd(),
    ignoreFile = ".vscodeignore",
    dependencies = [],
    readme = "README.md",
  } = options;

  // TODO: fix all of this ignore file handling.
  const gitignorePath = path.join(cwd, ".gitignore");
  const vscodeIgnorePath = path.join(cwd, ignoreFile);

  const ig = ignore();

  if (existsSync(gitignorePath)) {
    const ignoreContent = await readFile(gitignorePath, "utf8");
    ig.add(ignoreContent);
  }

  if (existsSync(vscodeIgnorePath)) {
    const vsceIgnoreContent = await readFile(vscodeIgnorePath, "utf8");
    ig.add(vsceIgnoreContent);
  }

  const globbedFiles = await glob("**", {
    cwd,
    followSymbolicLinks: true,
    expandDirectories: true,
    ignore: [...VSCE_DEFAULT_IGNORE, "!package.json", `!${readme}`, "node_modules/**"],
    dot: true,
    onlyFiles: true,
  });

  const filteredFiles = globbedFiles.filter((file) => !ig.ignores(file));

  const files = filteredFiles.map((file) => ({
    type: "local",
    localPath: path.join(cwd, file),
    path: path.join("extension/", file),
  })) satisfies VsixFile[];

  if (dependencies.length > 0) {
    for (const dep of dependencies) {
      files.push({
        type: "local",
        path: path.join("extension/node_modules", dep.name),
        localPath: dep.path,
      });
    }
  }

  return files;
}

/**
 * Detects the package manager used in a specific directory.
 * 
 * @param cwd - The current working directory to analyze for package manager detection
 * @returns The name of the detected package manager
 * @throws An error if no package manager is detected or if an unsupported package manager is found
 * 
 * @remarks
 * This function supports npm, yarn, and pnpm package managers. It will throw an error for deno and bun.
 */
export async function getExtensionPackageManager(cwd: string): Promise<PackageManager | null> {
  const result = await detect({ cwd });

  if (result == null) {
    throw new Error("could not detect package manager");
  }

  if (result.name === "deno" || result.name === "bun") {
    throw new Error(`unsupported package manager: ${result.name}`);
  }

  return result.name;
}

// taken from https://github.com/microsoft/vscode-vsce/blob/06951d9f03b90947df6d5ad7d9113f529321df20/src/package.ts#L1581-L1584
// added more default mime types to speed up the process
const DEFAULT_MIME_TYPES = new Map<string, string>([
  [".json", "application/json"],
  [".vsixmanifest", "text/xml"],
  [".md", "text/markdown"],
  [".png", "image/png"],
  [".txt", "text/plain"],
  [".js", "application/javascript"],
  [".yml", "text/yaml"],
  [".html", "text/html"],
  [".markdown", "text/markdown"],
  [".css", "text/css"],
]);

export interface ContentTypeResult {
  /**
   * The Content Types as a map of file extensions to content types.
   */
  contentTypes: Record<string, string>;

  /**
   * The Content Types as an XML string.
   *
   * NOTE:
   * Used inside the VSIX package to define the content types of the files.
   */
  xml: string;
}

/**
 * Generates content types mapping and XML representation for VSIX files.
 *
 * @param files - Array of VSIX files to process
 * @returns Object containing XML string representation of content types and mapping of file extensions to content types
 * @throws Error when content type cannot be determined for a file
 *
 * @remarks
 * This function determines the MIME content type for each file extension in the provided files.
 * It uses a default set of MIME types and falls back to mime type detection when necessary.
 *
 * @example
 * const files = [
 *   { type: "local", path: "extension/package.json", localPath: "/path/to/extension/package.json" },
 *   { type: "local", path: "extension/README.md", localPath: "/path/to/extension/README.md" }
 * ];
 * const { xml, contentTypes } = getContentTypesForFiles(files);
 */
export function getContentTypesForFiles(files: VsixFile[]): ContentTypeResult {
  const contentTypes: Record<string, string> = {};
  for (const file of files) {
    const ext = path.extname(file.path).toLowerCase();

    if (ext == null) continue;

    if (!DEFAULT_MIME_TYPES.has(ext)) {
      // if default mime types doesn't contain ext, lookup the mime type
      const contentType = mime.getType(ext);

      if (contentType == null) {
        throw new Error(`could not determine content type for file: ${file.path}`);
      }

      contentTypes[ext] = contentType;
    } else {
      contentTypes[ext] = DEFAULT_MIME_TYPES.get(ext)!;
    }
  }

  const xml = /* xml */`<?xml version="1.0" encoding="utf-8"?>
            <Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
              ${Object.entries(contentTypes).map(([ext, contentType]) => `<Default Extension="${ext}" ContentType="${contentType}" />\n`).join("")}
             </Types>
          `;

  return {
    xml,
    contentTypes,
  };
}

export interface TransformedFiles {
  /**
   * The assets to include in the manifest.
   *
   * NOTE:
   * These assets are not the extension file, but other files like changelog, readme, license, etc.
   */
  assets: ManifestAsset[];

  /**
   * The icon file path.
   */
  icon?: string;

  /**
   * The license file path.
   */
  license?: string;
}

export interface TransformFilesOptions {
  /**
   * The manifest object containing package details.
   */
  manifest: Manifest;

  /**
   * Files to process.
   */
  files: VsixFile[];

  /**
   * README file path
   */
  readme?: string;
}

/**
 * Transforms files for a VSIX package by identifying and categorizing specific asset files.
 *
 * @remarks
 * This function processes a collection of files to determine which assets should be included in the VSIX manifest.
 * It looks for specific files like license, icon, README, changelog, and translation files.
 *
 * @param options - Configuration options for file transformation
 * @param options.manifest - The extension manifest
 * @param options.files - Array of files to process
 * @param options.readme - Optional README file path
 *
 * @returns A promise resolving to transformed files metadata
 *
 * @throws {Error} If a license file is found but cannot be located in the files array
 */
export async function transformFiles(options: TransformFilesOptions): Promise<TransformedFiles> {
  const { manifest, files, readme } = options;

  const assets: ManifestAsset[] = [];
  let license: string | undefined;
  let icon: string | undefined;

  const hasLicenseFile = hasExtensionFile(files, ["LICENSE", "LICENSE.md", "LICENSE.txt", "LICENSE.markdown"]);
  const hasIconFile = hasExtensionFile(files, [manifest.icon]);
  // TODO: use default readme file names
  const hasReadmeFile = hasExtensionFile(files, [readme, "README.md"]);
  const hasChangelogFile = hasExtensionFile(files, ["CHANGELOG.md", "CHANGELOG.markdown", "CHANGELOG.txt"]);
  const hasTranslationsFiles = hasExtensionFile(files, ["package.nls.json"]);

  if (hasLicenseFile.found) {
    if (!extname(hasLicenseFile.path!)) {
      const entryIndex = files.findIndex((f) => f.path === hasLicenseFile.path);

      if (entryIndex === -1) {
        throw new Error(`could not find license file: ${hasLicenseFile.path}`);
      }

      const entry = files[entryIndex!]!;

      files[entryIndex!] = {
        ...entry,
        path: `${hasLicenseFile.path}.md`,
      };

      hasLicenseFile.path = files[entryIndex!]!.path;
    }

    license = hasLicenseFile.path;

    assets.push({
      type: "Microsoft.VisualStudio.Services.Content.License",
      path: license!,
    });
  }

  if (hasIconFile.found) {
    icon = hasIconFile.path;

    assets.push({
      type: "Microsoft.VisualStudio.Services.Icons.Default",
      path: icon!,
    });
  }

  if (hasReadmeFile.found) {
    assets.push({
      type: "Microsoft.VisualStudio.Services.Content.Details",
      path: hasReadmeFile.path!,
    });
  }

  if (hasChangelogFile.found) {
    assets.push({
      type: "Microsoft.VisualStudio.Services.Content.Changelog",
      path: hasChangelogFile.path!,
    });
  }

  if (hasTranslationsFiles.found) {
    assets.push({
      type: "Microsoft.VisualStudio.Code.Translation.en",
      path: hasTranslationsFiles.path!,
    });
  }

  return {
    assets,
    icon,
    license,
  };
}

function hasExtensionFile(files: VsixFile[], fileNames: (string | undefined)[]): { found: boolean; path: string | undefined } {
  for (const fileName of fileNames) {
    if (fileName == null) continue;
    const file = files.find((f) => f.path.endsWith(fileName));

    if (file) {
      return { found: true, path: file.path };
    }
  }

  return { found: false, path: undefined };
}
