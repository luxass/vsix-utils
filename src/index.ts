/**
 * @module
 *
 * This module contains utility functions for handling files.
 */

export type {
  CollectOptions,
  ContentTypeResult,
  ExtensionDependenciesOptions,
  ExtensionDependency,
  ProcessedFiles,
  VsixFile,
  VsixInMemoryFile,
  VsixLocalFile,
} from "./files";
export {
  collect,
  getContentTypesForFiles,
  getExtensionDependencies,
  getExtensionPackageManager,
  isInMemoryFile,
  isLocalFile,
  processFiles,
} from "./files";

export type {
  ManifestAsset,
  ManifestAssetType,
  ProjectManifest,
  VsixManifestOptions,
} from "./manifest";

export {
  createVsixManifest,
  getManifestTags,
  isWebKind,
  readProjectManifest,
  transformExtensionKind,
} from "./manifest";
export { prepublish, type PrepublishOptions } from "./scripts";
export type {
  Manifest,
  PackageManager,
  PackageManagerWithAuto,
} from "./types";
export type { ManifestValidation } from "./validation";
export {
  ALLOWED_SPONSOR_PROTOCOLS,
  EXTENSION_NAME_REGEX,
  EXTENSION_PRICING,
  GITHUB_BADGE_URL_REGEX,
  VALID_EXTENSION_KINDS,
  validateProjectManifest,
  validateVSCodeTypesCompatability,
  VSCODE_ENGINE_COMPATIBILITY_REGEX,
} from "./validation";

export type {
  RawVsixPackage,
  ReadVsixOptions,
  WriteVsixOptions,
} from "./zip";

export {
  readVsix,
  writeVsix,
} from "./zip";
