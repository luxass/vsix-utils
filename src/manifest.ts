/**
 * This module contains utility functions for manifests.
 * @module manifest
 *
 * @example
 * ```ts
 * import { readProjectManifest } from "vsix-utils";
 *
 * const projectManifest = await readProjectManifest('/path/to/project');
 *
 * if (projectManifest != null) {
 *   console.log("an error occurred while reading the project manifest");
 * }
 *
 * const { fileName, manifest } = projectManifest;
 *
 * console.log(fileName); // Outputs: /path/to/project/package.json
 * console.log(manifest); // Outputs: Parsed content of package.json
 * ```
 */

import type { Manifest } from "./types";
import { readFile } from "node:fs/promises";
import path from "node:path";

export interface ProjectManifest {
  fileName: string;
  manifest: Manifest;
}

/**
 * Reads the project manifest (package.json) from the specified project directory.
 *
 * @param {string} projectDir - The directory of the project where the package.json is located.
 * @returns {Promise<ProjectManifest | null>} A promise that resolves to an object containing the file name and the parsed manifest.
 *
 * @example
 * ```ts
 * const { fileName, manifest } = await readProjectManifest('/path/to/project');
 * console.log(fileName); // Outputs: /path/to/project/package.json
 * console.log(manifest); // Outputs: Parsed content of package.json
 * ```
 */
export async function readProjectManifest(projectDir: string): Promise<ProjectManifest | null> {
  try {
    const manifestPath = path.join(projectDir, "package.json");
    const manifest = JSON.parse(await readFile(manifestPath, "utf-8"));

    return {
      fileName: manifestPath,
      manifest,
    };
  } catch {
    return null;
  }
}
