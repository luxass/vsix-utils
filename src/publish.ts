/**
 * @module publish
 *
 * This module contains utility functions for publishing vsix files.
 */

import type { Manifest } from "./types";
import type { VSCE_TARGETS } from "./vsce-constants";

export interface PublishOptions {
  personalAccessToken?: string;

  targets?: typeof VSCE_TARGETS[];

  packagePaths?: string[];

  /**
   * Additional headers to apply to the request
   */
  additonalHeaders?: Record<string, string>;

  /**
   * Which registry to publish extensions to
   * @default "vsmarketplace"
   *
   * NOTE:
   * You can pass a object with registryUrl, if you are using a custom registry.
   */
  registry?: "vsmarketplace" | "openvsx" | { registryUrl: string };
}

export async function publish(options: PublishOptions): Promise<any> {
  const { registry = "vsmarketplace", additonalHeaders, packagePaths, targets, personalAccessToken } = options;

  const registryUrl = (typeof registry === "object" && "registryUrl" in registry) ? registry.registryUrl : registry === "openvsx" ? "https://open-vsx.org" : "https://marketplace.visualstudio.com";

  const publishRequests = [];
  // support multiple targets combined with multiple package paths
  // support multiple package paths without targets (auto detect target)
}
