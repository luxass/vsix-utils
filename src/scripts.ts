import type { Manifest, PackageManager } from "./types";
import { spawn } from "node:child_process";
import { promisify } from "node:util";

const spawnAsync = promisify(spawn);

export interface PrepublishOptions {
  /**
   * The current working directory.
   */
  cwd: string;

  /**
   * The package manager to use.
   */
  packageManager: Exclude<PackageManager, "auto">;

  /**
   * The manifest of the extension.
   */
  manifest: Manifest;

  /**
   * Whether the extension is a pre-release.
   */
  preRelease: boolean;
}

export async function prepublish(options: PrepublishOptions): Promise<void> {
  const { cwd, packageManager, manifest, preRelease } = options;

  if (!manifest.scripts) {
    return;
  }

  const scripts: string[] = [];
  if ("vscode:prepublish" in manifest.scripts) {
    scripts.push("vscode:prepublish");
  }

  if ("vscode:prepublish:release" in manifest.scripts && !preRelease) {
    scripts.push("vscode:prepublish:release");
  }

  if ("vscode:prepublish:prerelease" in manifest.scripts && preRelease) {
    scripts.push("vscode:prepublish:prerelease");
  }

  if (scripts.length === 0) {
    return;
  }

  for (const script of scripts) {
    await runScript({ cwd, packageManager, script });
  }
}

export interface RunScriptOptions {
  /**
   * The current working directory.
   */
  cwd: string;

  /**
   * The package manager to use.
   */
  packageManager: Exclude<PackageManager, "auto">;

  /**
   * The script to run.
   */
  script: string;
}

export async function runScript(options: RunScriptOptions): Promise<void> {
  const { cwd, packageManager, script } = options;

  await spawnAsync(packageManager, ["run", script], {
    cwd,
    shell: true,
    stdio: "inherit",
  });
}
