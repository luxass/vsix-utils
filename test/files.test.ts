import { readFile } from "node:fs/promises";
import { join, normalize } from "node:path";

import { assert, describe, expect, it } from "vitest";
import { fromFileSystem, testdir } from "vitest-testdirs";
import { collect, getContentTypesForFiles, type VsixFile } from "../src/files";
import { readProjectManifest } from "../src/manifest";

it("should collect files for a simple extension", async () => {
  const testdirFiles = await fromFileSystem("./test/fixtures/extensions/simple-extension");
  const path = await testdir(testdirFiles);

  const projectManifest = await readProjectManifest(path);

  if (projectManifest == null) {
    expect.fail("project manifest is null");
  }

  const { manifest } = projectManifest;

  const files = await collect(manifest, {
    cwd: path,
  });

  assert(manifest.displayName === "Simple Extension", "displayName should be 'Simple Extension'");
  assert(files.length > 0, "files should not be empty");

  expect(files).toMatchObject(expect.arrayContaining([
    {
      type: "local",
      path: normalize("extension/package.json"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-simple-extension/package.json"),
    },
    {
      type: "local",
      path: normalize("extension/README.md"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-simple-extension/README.md"),
    },
    {
      type: "local",
      path: normalize("extension/LICENSE"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-simple-extension/LICENSE"),
    },
  ] satisfies VsixFile[]));
});

it("should collect files for a extension with a different README path", async () => {
  const testdirFiles = await fromFileSystem("./test/fixtures/extensions/simple-extension", {
    extras: {
      "extra-readme.md": "This is an extra README file",
    },
  });

  const path = await testdir(testdirFiles);

  const projectManifest = await readProjectManifest(path);

  if (projectManifest == null) {
    expect.fail("project manifest is null");
  }

  const { manifest } = projectManifest;

  const files = await collect(manifest, {
    cwd: path,
    readme: "extra-readme.md",
  });

  const readmeContent = await readFile(join(path, "extra-readme.md"), "utf8");

  assert(manifest.displayName === "Simple Extension", "displayName should be 'Simple Extension'");
  assert(files.length > 0, "files should not be empty");

  expect(files).toMatchObject(expect.arrayContaining([
    {
      type: "local",
      path: normalize("extension/package.json"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-README-path/package.json"),
    },
    {
      type: "local",
      path: normalize("extension/README.md"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-README-path/README.md"),
    },
    {
      type: "local",
      path: normalize("extension/LICENSE"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-README-path/LICENSE"),
    },
  ] satisfies VsixFile[]));

  expect(readmeContent).toBe("This is an extra README file");
});

it("should collect files for a extension with a different ignore file", async () => {
  const testdirFiles = await fromFileSystem("./test/fixtures/extensions/simple-extension", {
    extras: {
      ".vsixignore": ".vscode/**\n.gitignore\n**/tsconfig.json",
    },
  });

  const path = await testdir(testdirFiles);

  const projectManifest = await readProjectManifest(path);

  if (projectManifest == null) {
    expect.fail("project manifest is null");
  }

  const { manifest } = projectManifest;

  const files = await collect(manifest, {
    cwd: path,
    ignoreFile: ".vsixignore",
  });

  assert(manifest.displayName === "Simple Extension", "displayName should be 'Simple Extension'");
  assert(files.length > 0, "files should not be empty");

  expect(files).toMatchObject(expect.arrayContaining([
    {
      type: "local",
      path: normalize("extension/package.json"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-ignore-file/package.json"),
    },
    {
      type: "local",
      path: normalize("extension/README.md"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-ignore-file/README.md"),
    },
    {
      type: "local",
      path: normalize("extension/LICENSE"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-ignore-file/LICENSE"),
    },
    {
      type: "local",
      path: normalize("extension/src/extension.ts"),
      localPath: normalize(".vitest-testdirs/vitest-files-should-collect-files-for-a-extension-with-a-different-ignore-file/src/extension.ts"),
    },
  ] satisfies VsixFile[]));
});

describe("content types", () => {
  it("should handle empty input", () => {
    const result = getContentTypesForFiles([]);

    expect(result.contentTypes).toEqual({});
    expect(result.xml).not.toContain("<Default");
  });

  it("should assign default content types for known extensions", () => {
    const result = getContentTypesForFiles([
      {
        type: "local",
        path: "file.txt",
        localPath: "file.txt",
      },
      {
        type: "local",
        path: "file.json",
        localPath: "file.json",
      },
    ]);

    expect(result.contentTypes).toEqual({
      ".txt": "text/plain",
      ".json": "application/json",
    });

    expect(result.xml).toContain("<Default Extension=\".txt\" ContentType=\"text/plain\" />");
    expect(result.xml).toContain("<Default Extension=\".json\" ContentType=\"application/json\" />");
    expect(result.xml).toMatchSnapshot();
  });

  it("should lookup content types for unknown extensions", () => {
    const result = getContentTypesForFiles([
      {
        type: "local",
        path: "file.jade",
        localPath: "file.jade",
      },
    ]);

    expect(result.contentTypes).toEqual({
      ".jade": "text/jade",
    });

    expect(result.xml).toContain("<Default Extension=\".jade\" ContentType=\"text/jade\" />");
    expect(result.xml).toMatchSnapshot();
  });

  it("should throw an error for unknown extensions that cannot be resolved", () => {
    expect(() => getContentTypesForFiles([
      {
        type: "local",
        localPath: "file.unknown",
        path: "file.unknown",
      },
    ])).toThrow(
      "could not determine content type for file: file.unknown",
    );
  });
});
