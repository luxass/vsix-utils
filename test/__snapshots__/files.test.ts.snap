// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`content types > should assign default content types for known extensions 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
            <Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
              <Default Extension=".txt" ContentType="text/plain" />
<Default Extension=".json" ContentType="application/json" />

             </Types>
          "
`;

exports[`content types > should lookup content types for unknown extensions 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
            <Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
              <Default Extension=".jade" ContentType="text/jade" />

             </Types>
          "
`;
